tests/h/settings_test.py

```diff
@@ -3,98 +3,104 @@
-                                          message='what to do instead')
-        func.warn = mock.Mock(spec_set=[])
-
-        result = func({})
-
-        assert result == {'foo': 'bar'}
-        func.warn.assert_called_once_with('use of fake setting is '
-                                          'deprecated: what to do instead')
+def test_database_url():
+    url = 'postgres://postgres:1234/database'
+    expected = 'postgresql+psycopg2://postgres:1234/database'
 
-    def test_emits_no_warnings_when_unused(self):
-        func = settings.DeprecatedSetting(FakeSetting(result=None),
-                                          message='what to do instead')
-        func.warn = mock.Mock(spec_set=[])
+    assert settings.database_url(url) == expected
 
-        result = func({})
 
-        assert result is None
-        assert not func.warn.called
+class TestSettingsManager(object):
+    def test_does_not_warn_when_deprecated_setting_is_not_used(self, caplog):
+        with caplog.at_level(logging.WARN):
+            settings_manager = settings.SettingsManager({}, {})
+            settings_manager.set('foo', 'FOO', deprecated_msg='what to do instead')
+        assert not caplog.records
 
+    def test_sets_value_when_deprecated_setting_is_used(self):
+        settings_manager = settings.SettingsManager({}, {'FOO': 'bar'})
+        settings_manager.set('foo', 'FOO', deprecated_msg='what to do instead')
 
-def asutf8(str):
-    return str.encode()
+        result = settings_manager.settings['foo']
 
+        assert result == 'bar'
 
-@pytest.mark.parametrize('setting,varname,type,environ,expected', (
-    # Should return None when the env var in question isn't set
-    ('foo', 'FOO', None, {}, None),
+    def test_warns_when_deprecated_setting_is_used(self, caplog):
+        with caplog.at_level(logging.WARN):
+            settings_manager = settings.SettingsManager({}, {'FOO': 'bar'})
+            settings_manager.set('foo', 'FOO', deprecated_msg='what to do instead')
+        assert 'what to do instead' in caplog.text
 
-    # Should return the setting as a string when the env var is set
-    ('foo', 'FOO', None, {'FOO': 'bar'}, {'foo': 'bar'}),
-    ('foo.bar', 'FOO', None, {'FOO': 'baz'}, {'foo.bar': 'baz'}),
+    @pytest.mark.parametrize('config,setting,varname,type_,environ,required,default,expected', (
+        # Should leave value at None when the env var in question isn't set
+        ({'foo': None}, 'foo', 'FOO', str, {}, False, None, None),
 
-    # Should coerce the result using the passed type
-    ('foo', 'FOO', asutf8, {'FOO': 'bar'}, {'foo': b'bar'}),
-    ('app_port', 'PORT', int, {'PORT': '123'}, {'app_port': 123}),
-))
-def test_env_setting(setting, varname, type, environ, expected):
-    func = settings.EnvSetting(setting, varname, type)
+        # Should return the setting as a string when the env var is set
+        ({}, 'foo', 'FOO', str, {'FOO': 'bar'}, False, None, 'bar'),
```

Comment:
I find these long tuples really hard to read (trying to see which values go with which parameters of the test method below), it's too many params to be readable. Difficult for me to read this and try to figure out whether all the possible cases have been tested. What about you?

I wonder if to make it easier to read we should parametrize this test with a dict like:

```python
@pytest.mark.parametrize('params', [
    dict(settings={"foo": None}, name="foo", envvar="FOO", type_=str, environ={}, required=False, default=None, expected=None),
    dict(...),
    dict(...),
    ...
])
```

Then the test would receive a single `params` dict that it'd have to read the individual params from by name, instead of receiving separate arguments.